waspc/src/Wasp/AppSpec/Valid.hs

```diff
@@ -50,52 +51,71 @@ validateExactlyOneAppExists spec =
-      ]
-    else []
+  [ GenericValidationError
+      "Expected app.auth to be defined since there are Pages with authRequired set to true."
+    | anyPageRequiresAuth && not (isAuthEnabled spec)
+  ]
   where
     anyPageRequiresAuth = any ((== Just True) . Page.authRequired) (snd <$> AS.getPages spec)
 
 validateDbIsPostgresIfPgBossUsed :: AppSpec -> [ValidationError]
 validateDbIsPostgresIfPgBossUsed spec =
-  if isPgBossJobExecutorUsed spec && not (isPostgresUsed spec)
-    then
-      [ GenericValidationError
-          "Expected app.db.system to be PostgreSQL since there are jobs with executor set to PgBoss."
-      ]
-    else []
+  [ GenericValidationError
+      "Expected app.db.system to be PostgreSQL since there are jobs with executor set to PgBoss."
+    | isPgBossJobExecutorUsed spec && not (isPostgresUsed spec)
+  ]
 
 validateAuthUserEntityHasCorrectFieldsIfUsernameAndPasswordAuthIsUsed :: AppSpec -> [ValidationError]
 validateAuthUserEntityHasCorrectFieldsIfUsernameAndPasswordAuthIsUsed spec = case App.auth (snd $ getApp spec) of
   Nothing -> []
   Just auth ->
-    if Auth.UsernameAndPassword `notElem` Auth.methods auth
+    if not $ Auth.isUsernameAndPasswordAuthEnabled auth
       then []
       else
         let userEntity = snd $ AS.resolveRef spec (Auth.userEntity auth)
             userEntityFields = Entity.getFields userEntity
-            maybeUsernameField = find ((== "username") . Entity.Field.fieldName) userEntityFields
-            maybePasswordField = find ((== "password") . Entity.Field.fieldName) userEntityFields
-         in concat
-              [ case maybeUsernameField of
-                  Just usernameField
-                    | Entity.Field.fieldType usernameField == Entity.Field.FieldTypeScalar Entity.Field.String -> []
-                  _ ->
-                    [ GenericValidationError
-                        "Expected an Entity referenced by app.auth.userEntity to have field 'username' of type 'string'."
-                    ],
-                case maybePasswordField of
-                  Just passwordField
-                    | Entity.Field.fieldType passwordField == Entity.Field.FieldTypeScalar Entity.Field.String -> []
-                  _ ->
-                    [ GenericValidationError
-                        "Expected an Entity referenced by app.auth.userEntity to have field 'password' of type 'string'."
-                    ]
+         in concatMap
+              (validateField "app.auth.userEntity" userEntityFields)
+              [ ("username", Entity.Field.FieldTypeScalar Entity.Field.String, "String"),
+                ("password", Entity.Field.FieldTypeScalar Entity.Field.String, "String")
               ]
 
+validateExternalAuthAssociationEntityHasCorrectFieldsIfExternalAuthIsUsed :: AppSpec -> [ValidationError]
+validateExternalAuthAssociationEntityHasCorrectFieldsIfExternalAuthIsUsed spec = case App.auth (snd $ getApp spec) of
```

Comment:
Actually, you validate it has correct fields even if external auth is not used, right? Because in that case you return [] which is correct? Maybe then you don't have to mention that in the method name? Or you wanted to make it emphasized that it is checking for this stuff?

Reply:
I can see what you mean. I mainly kept with the convention of the existing function, `validateAuthUserEntityHasCorrectFieldsIfUsernameAndPasswordAuthIsUsed`. I think its prob fine either way, but can change if anyone feels strongly.